Assets/MixedRealityToolkit-SDK/Features/UX/Scripts/Collections/GridObjectCollection.cs

```diff
@@ -0,0 +1,446 @@
+
+        }
+
+        /// <summary>
+        /// This is the radius of either the Cylinder or Sphere mapping and is ignored when using the plane mapping.
+        /// </summary>
+        public float Radius
+        {
+            get { return radius; }
+            set { radius = value; }
+
+        }
+
+        /// <summary>
+        /// This is the radial range for creating a radial fan layout.
+        /// </summary>
+        public float RadialRange
+        {
+            get { return radialRange; }
+            set { radialRange = value; }
+        }
+
+        /// <summary>
+        /// Number of rows per column, column number is automatically determined
+        /// </summary>
+        public int Rows
+        {
+            get { return rows; }
+            set { rows = value; }
+
+        }
+
+        /// <summary>
+        /// Width of the cell per object in the collection.
+        /// </summary>
+        public float CellWidth
+        {
+            get { return cellWidth; }
+            set { cellWidth = value; }
+
+        }
+
+        /// <summary>
+        /// Height of the cell per object in the collection.
+        /// </summary>
+        public float CellHeight
+        {
+            get { return cellHeight; }
+            set { cellHeight = value; }
+
+        }
+
+        /// <summary>
+        /// Reference mesh to use for rendering the sphere layout
+        /// </summary>
+        public Mesh SphereMesh
+        {
+            get { return sphereMesh; }
+            set { sphereMesh = value; }
+
+        }
+
+        /// <summary>
+        /// Reference mesh to use for rendering the cylinder layout
+        /// </summary>
+        public Mesh CylinderMesh
+        {
+            get { return cylinderMesh; }
+            set { cylinderMesh = value; }
+        }
+
+        public float Width
+        {
+            get { return width; }
+        }
+
+        public float Height
+        {
+            get { return height; }
+        }
+        #endregion public accessors
+
+
+        /// <summary>
+        /// Overriding base function function for laying out all the children when UpdateCollection is called.
+        /// </summary>
+        protected override void LayoutChildren()
+        {
+            float startOffsetX;
+            float startOffsetY;
+
+            Vector3[] nodeGrid = new Vector3[NodeList.Count];
+            Vector3 newPos = Vector3.zero;
+
+            // Now lets lay out the grid
+            startOffsetX = (columns * 0.5f) * CellWidth;
+            startOffsetY = (rows * 0.5f) * CellHeight;
+            halfCell = new Vector2(CellWidth * 0.5f, CellHeight * 0.5f);
+
+            // First start with a grid then project onto surface
+            ResolveGridLayout(nodeGrid, startOffsetX, startOffsetY, layout);
+
+            switch (SurfaceType)
+            {
+                case ObjectOrientationSurfaceTypeEnum.Plane:
+                    for (int i = 0; i < NodeList.Count; i++)
```

Comment:
Would this inheritance be more useful if each SurfaceType called an overridable function? That way you would just override the behaviors you wanted differently, and could leave the rest of the LayoutChildren() function alone?

Reply:
Just seems like this is a huge block of functionality to override. :) (I am lazy)

Reply:
That assumes SurfaceType needs to be implemented - which it doesn't. This Collection just happens to implement LayoutChildren with Surface types. The BaseObjectCollection contains the abstract LayoutChildren method - so more likely you would be inheriting the base collection and writing from scratch - not the grid collection. This is actually the main update I wanted to get into MRTK, as it is, the ObjectCollection is a great control, but since all of its specific functionality is embedded in the single class, its not very extensible.

Reply:
Aha! This all click into sense now that we've talked!